hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java

```diff
@@ -18,27 +18,139 @@
+            @NonNull final HandleContext context,
+            @NonNull final UtilPrngTransactionBody op,
+            @NonNull final PrngConfig prngConfig,
+            @NonNull final PrngRecordBuilder recordBuilder) {
+        final var range = op.range();
+
+        // TODO: This check should probably be moved into app
+        if (!prngConfig.isPrngEnabled()) {
+            return;
+        }
+        // get the n-3 running hash. If the running hash is not available, will throw a
+        // HandleException
+        final var runningHashStore = context.createStore(ReadableRunningHashLeafStore.class);
+        final var nMinusThreeRunningHash = runningHashStore.getNMinusThreeRunningHash();
+        final byte[] pseudoRandomBytes = getNMinus3RunningHashBytes(nMinusThreeRunningHash);
+
+        // If no bytes are available then return
+        if (pseudoRandomBytes == null || pseudoRandomBytes.length == 0) {
+            return;
+        }
+        // If range is provided then generate a random number in the given range
+        // from the pseudoRandomBytes
+        if (range > 0) {
+            final int pseudoRandomNumber = randomNumFromBytes(pseudoRandomBytes, range);
+            recordBuilder.setPrngNumber(pseudoRandomNumber);
+        } else {
+            recordBuilder.setPrngBytes(Bytes.wrap(pseudoRandomBytes));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PrngRecordBuilder newRecordBuilder() {
+        return new UtilPrngRecordBuilder();
+    }
+
+    /**
+     * Get the n-3 running hash bytes from the n-3 running hash. If the running hash is not
+     * available, will throw a HandleException. n-3 running hash is chosen for generating random
+     * number instead of n-1 running hash for processing transactions quickly. Because n-1 running
+     * hash might not be available when the transaction is processed and might need to wait longer.
+     *
+     * @param nMinus3RunningHash n-3 running hash
+     * @return n-3 running hash bytes
+     */
+    private byte[] getNMinus3RunningHashBytes(@NonNull final RunningHash nMinus3RunningHash) {
+        // This can't happen because this running hash is taken from record stream.
+        // If this happens then there is a bug in the code.
+        requireNonNull(nMinus3RunningHash);
+
+        Hash nMinusThreeHash;
+        try {
+            nMinusThreeHash = nMinus3RunningHash.getFutureHash().get();
+            // Use n-3 running hash instead of n-1 running hash for processing transactions quickly
+            if (nMinusThreeHash == null || Arrays.equals(nMinusThreeHash.getValue(), new byte[48])) {
+                log.info("No n-3 record running hash available to generate random number");
+                return MISSING_BYTES;
+            }
+            // generate binary string from the running hash of records
+            return nMinusThreeHash.getValue();
+        } catch (InterruptedException | ExecutionException e) {
```

Comment:
We probably shouldn't try to combine these two.  Interrupted means the process is being asked to exit "cleanly".  ExecutionException is just any exception (runtime, checked, etc..) that occurred on the other thread.  The first should interrupt, but I don't thing ExecutionException should do that.
Also looks like a partial mingle at the bottom with a TODO after throwing an exception.

Reply:
Separated both exceptions. Added `UNKNOWN` response, which might be changed in the future